middleware/rate_limiter.go

```diff
@@ -0,0 +1,223 @@
+		RateLimiterMemoryStoreRate(1),
+		RateLimiterMemoryStoreBurst(3),
+		RateLimiterMemoryStoreExpiresIn(3 * time.Minute),
+	)
+
+	config := RateLimiterConfig{
+		Skipper: DefaultSkipper,
+		IdentifierExtractor: func(ctx echo.Context) (string, error) {
+			id := ctx.RealIP()
+			return id, nil
+		},
+		ErrorHandler: func(context echo.Context) error {
+			return context.JSON(http.StatusTooManyRequests, nil)
+		},
+		DenyHandler: func(context echo.Context) error {
+			return context.JSON(http.StatusForbidden, nil)
+		},
+	}
+
+	e.GET("/rate-limited", func(c echo.Context) error {
+		return c.String(http.StatusOK, "test")
+	}, RateLimiterWithConfig(config))
+*/
+func RateLimiterWithConfig(config RateLimiterConfig) echo.MiddlewareFunc {
+	if config.Skipper == nil {
+		config.Skipper = DefaultRateLimiterConfig.Skipper
+	}
+	if config.IdentifierExtractor == nil {
+		config.IdentifierExtractor = DefaultRateLimiterConfig.IdentifierExtractor
+	}
+	if config.ErrorHandler == nil {
+		config.ErrorHandler = DefaultRateLimiterConfig.ErrorHandler
+	}
+	if config.DenyHandler == nil {
+		config.DenyHandler = DefaultRateLimiterConfig.DenyHandler
+	}
+	if config.Store == nil {
+		panic("Store configuration must be provided")
+	}
+	return func(next echo.HandlerFunc) echo.HandlerFunc {
+		return func(c echo.Context) error {
+			if config.Skipper(c) {
+				return next(c)
+			}
+			if config.BeforeFunc != nil {
+				config.BeforeFunc(c)
+			}
+
+			identifier, err := config.IdentifierExtractor(c)
+			if err != nil {
+				return config.DenyHandler(c)
+			}
+
+			if !config.Store.Allow(identifier) {
+				return config.ErrorHandler(c)
+			}
+			return next(c)
+		}
+	}
+}
+
+type (
+	// RateLimiterMemoryStore is the built-in store implementation for RateLimiter
+	RateLimiterMemoryStore struct {
+		visitors    map[string]*Visitor
+		mutex       sync.Mutex
+		rate        rate.Limit
+		burst       int
+		expiresIn   time.Duration
+		lastCleanup time.Time
+	}
+	// Visitor signifies a unique user's limiter details
+	Visitor struct {
+		*rate.Limiter
+		lastSeen time.Time
+	}
+)
+
+// NewRateLimiterMemoryStore returns an instance of RateLimiterMemoryStore
+func NewRateLimiterMemoryStore(config RateLimiterMemoryStoreConfig) (store *RateLimiterMemoryStore) {
+	store = &RateLimiterMemoryStore{}
+
+	store.rate = config.rate
+	store.burst = config.burst
+	if config.expiresIn == 0 {
+		store.expiresIn = DefaultRateLimiterMemoryStoreConfig.expiresIn
+	} else {
+		store.expiresIn = config.expiresIn
+	}
+	store.visitors = make(map[string]*Visitor)
+	store.lastCleanup = now()
+	return
+}
+
+// RateLimiterMemoryStoreConfig represents configuration for RateLimiterMemoryStore
+type RateLimiterMemoryStoreConfig struct {
+	rate      rate.Limit
+	burst     int
+	expiresIn time.Duration
+}
+
+// DefaultRateLimiterMemoryStoreConfig provides default configuration values for RateLimiterMemoryStore
+var DefaultRateLimiterMemoryStoreConfig = RateLimiterMemoryStoreConfig{
```

Comment:
It will be hard to choose a good default probably, as it highly depends on the traffic and use case.
Fail fast seems the easiest approach, by telling the user we fail if the rate is 0 (in which case the limiter does not make sense) and we should bail out anyway.

I also suggested using a simple `NewRateLimiterMemoryStore(rate int)` (don't know if burst is needed too) and a `RateLimiterMemoryStoreWithConfig(RateLimiterMemoryStoreConfig{})` (although a little long) would give the best of both worlds probably.

Side note: We cannot use `RateLimiterMemoryStore` as it is the name of the struct, so `NewRateLimiterMemoryStore` can be used.

@iambenkay Sorry for being a little pedantic here, it is just an attempt to get it right.